### PR TITLE
Fix minor bugs in timeseries product codes

### PR DESCRIPTION
**Pull request by mhidas**
Just picking off some low-hanging fruit.

Addresses the first two bugs listed in https://github.com/aodn/PO-Backlog/issues/1027
I was hoping to tackle the third one as well, but it might be easier to do it in a separate PR.

### CODE REVIEW

**Comment by mhidas** [comment]
@ocehugo Ready for review. There's quite a bit going on. It might make more sense looking at it one commit at a time (hopefully).

Note: I know the unittests are not quite complete, and there's plenty of refactoring that could be done, but these will be addressed later. For now I'm just trying to focus on fixing those bugs.

File: test_aodntools/timeseries_products/test_aggregated_timeseries.py [hunk 1]
```diff
@@ -60,6 +63,12 @@ def test_main_aggregator(self):
         self.assertIn(__version__, dataset.lineage)
         self.assertIn(BAD_FILE, dataset.rejected_files)
 
+        # check aggregated variable values
+        expected = Dataset(EXPECTED_OUTPUT_FILE)
+        non_match_vars = []
```
Comment (ocehugo) [inline test_aodntools/timeseries_products/test_aggregated_timeseries.py:74]: I dont think you need this anymore

File: test_aodntools/timeseries_products/test_velocity_aggregated_timeseries.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        for f in chartostring(dataset['source_file'][:]):
+            self.assertIn(f, INPUT_FILES)
+
+        # check attributes
+        self.assertEqual(__version__, dataset.generating_code_version)
+        self.assertIn(__version__, dataset.lineage)
+
+        # check aggregated variable values
+        expected = Dataset(EXPECTED_OUTPUT_FILE)
+        non_match_vars = []
```
Comment (ocehugo) [inline test_aodntools/timeseries_products/test_velocity_aggregated_timeseries.py:70]: as above

File: aodntools/timeseries_products/aggregated_timeseries.py [hunk 3]
```diff
@@ -267,39 +267,35 @@ def main_aggregator(files_to_agg, var_to_agg, site_code, input_dir='', output_di
...
+                n_obs += len(nc.TIME)
             else:
                 bad_files.update({file: error_list})
                 rejected_files.append(file)
 
-    ## remove bad files form the list
+    ## remove bad files form the list and sort in chronological order
     for file in bad_files.keys():
         files_to_agg.remove(file)
+    files_to_agg = sort_files(files_to_agg, input_dir=input_dir)
```
Comment (ocehugo) [inline aodntools/timeseries_products/aggregated_timeseries.py:304]: rename to sort_by_deployment_date

File: aodntools/timeseries_products/aggregated_timeseries.py [hunk 4]
```diff
@@ -331,26 +327,29 @@ def main_aggregator(files_to_agg, var_to_agg, site_code, input_dir='', output_di
...
 
     ## main loop
+    start = 0
     for index, file in enumerate(files_to_agg):
         with xr.open_dataset(os.path.join(input_dir, file)) as nc:
             nc = in_water(nc)
-            start = sum(varlen_list[:index + 1])
-            end = sum(varlen_list[:index + 2])
+            n_obs = len(nc.TIME)
+            end = start + n_obs
```
Comment (ocehugo) [inline aodntools/timeseries_products/aggregated_timeseries.py:347]: rename n_obs to total_nobs above, yo are mutating n_obs here. I would also use slices:
```
nc = in_water(nc)
chunk = slice(start,start+n_obs);
DEPTH[chunk] ...
...
start = n_obs;
```

In fact, this ```with``` block shoudl be a function receiving ```ds,dsvars,nc```, after you define some top-level constants (one_day,epoch,etc).

Comment (ocehugo) [reply aodntools/timeseries_products/aggregated_timeseries.py:347]: Also, you could store the ```len(nc.TIME)``` of the files from the loop above (278)) and use it here in the for iterator.

File: aodntools/timeseries_products/velocity_aggregated_timeseries.py [hunk 5]
```diff
@@ -121,36 +121,32 @@ def velocity_aggregated(files_to_agg, site_code, input_dir='', output_dir='./',
...
     bad_files = {}
 
     # default name for temporary file. It will be renamed at the end
     _, temp_outfile = tempfile.mkstemp(suffix='.nc', dir=output_dir)
 
-    ## sort the file list in chronological order
-    files_to_agg = utils.sort_files(files_to_agg, input_dir=input_dir)
-
     ## check files and get total number of flattened obs
+    n_obs = 0
```
Comment (ocehugo) [inline aodntools/timeseries_products/velocity_aggregated_timeseries.py:137]: repeating pattern - just factor out as a function to create a list of slices

File: aodntools/timeseries_products/velocity_aggregated_timeseries.py [hunk 6]
```diff
@@ -202,8 +197,8 @@ def velocity_aggregated(files_to_agg, site_code, input_dir='', output_dir='./',
                 WCUR[start:end] = flat_variable(nc, 'WCUR')
                 WCURqc[start:end] = flat_variable(nc, 'WCUR_quality_control')
             else:
-                WCUR[start:end] = np.full(varlen_list[index + 1], np.nan)
-                WCURqc[start:end] = np.full(varlen_list[index + 1], 9)
+                WCUR[start:end] = np.full(n_obs, np.nan)
```
Comment (ocehugo) [inline aodntools/timeseries_products/velocity_aggregated_timeseries.py:209]: same thing, mutating the value of n_obs.

File: aodntools/timeseries_products/hourly_timeseries.py [hunk 7]
```diff
@@ -540,7 +519,7 @@ def hourly_aggregator(files_to_aggregate, site_code, qcflags, input_dir='', outp
     nc_aggregated = nc_aggregated.drop('OBSERVATION')
 
     ## add global attributes
-    contributor_name, contributor_email, contributor_role = get_contributors(files_to_aggregate, input_dir=input_dir)
+    contributor_name, contributor_email, contributor_role = utils.get_contributors(files_to_aggregate, input_dir=input_dir)
```
Comment (ocehugo) [inline aodntools/timeseries_products/hourly_timeseries.py:525]: wrap that beast to return the add_attribute instead. ```get_contributor_attr```

File: aodntools/timeseries_products/velocity_hourly_timeseries.py [hunk 8]
```diff
@@ -121,6 +122,8 @@ def velocity_hourly_aggregated(files_to_agg, site_code, input_dir='', output_dir
     ## remove bad files form the list
     for file in bad_files.keys():
         files_to_agg.remove(file)
+    if len(files_to_agg) == 0:
```
Comment (ocehugo) [inline aodntools/timeseries_products/velocity_hourly_timeseries.py:125]: repeating pattern - factor out that file_to_agg removal/checker into a function.

File: aodntools/timeseries_products/hourly_timeseries.py (hunk 7 above)
Comment (mhidas) [reply aodntools/timeseries_products/hourly_timeseries.py:525]: Done.

File: aodntools/timeseries_products/aggregated_timeseries.py (hunk 4 above)
Comment (mhidas) [reply aodntools/timeseries_products/aggregated_timeseries.py:347]: I've renamed to `n_obs_total` above. The rest I don't really see the point, sorry. Anyway this is not meant to be a refactoring PR, so I'll try to keep these changes minimal.

File: aodntools/timeseries_products/velocity_hourly_timeseries.py (hunk 8 above)
Comment (mhidas) [reply aodntools/timeseries_products/velocity_hourly_timeseries.py:125]: Will do, later, in a refactoring PR.

File: test_aodntools/timeseries_products/test_aggregated_timeseries.py (hunk 1 above)
Comment (mhidas) [reply test_aodntools/timeseries_products/test_aggregated_timeseries.py]: I actually did mean to use this... updated.

**Comment by mhidas** [comment]
Thanks @ocehugo . I made a few simple changes based on your suggestions. As I said, I'd like to do more extensive refactoring in  a separate PR.

This should be good to merge now. Cheers

**Pull request merged**